source/extensions/filters/network/rbac/rbac_filter.cc

```diff
@@ -51,14 +53,28 @@ Network::FilterStatus RoleBasedAccessControlFilter::onData(Buffer::Instance&, bo
   return Network::FilterStatus::Continue;
 }
 
+void RoleBasedAccessControlFilter::setDynamicMetadata(std::string shadow_policy_id) {
+  ProtobufWkt::Struct metrics;
+  auto& fields = *metrics.mutable_fields();
+  if (!shadow_policy_id.empty()) {
+    *fields[Filters::Common::RBAC::shadow_policy_id_field].mutable_string_value() =
+        shadow_policy_id;
+  }
+  callbacks_->connection().streamInfo().setDynamicMetadata(NetworkFilterNames::get().Rbac, metrics);
+}
+
 EngineResult
 RoleBasedAccessControlFilter::checkEngine(Filters::Common::RBAC::EnforcementMode mode) {
   const auto& engine = config_->engine(mode);
   if (engine.has_value()) {
-    if (engine->allowed(callbacks_->connection())) {
+    std::string effective_policy_id;
+    if (engine->allowed(callbacks_->connection(),
+                        callbacks_->connection().streamInfo().dynamicMetadata(),
+                        &effective_policy_id)) {
       if (mode == Filters::Common::RBAC::EnforcementMode::Shadow) {
         ENVOY_LOG(debug, "shadow allowed");
         config_->stats().shadow_allowed_.inc();
+        setDynamicMetadata(effective_policy_id);
```

Comment:
1. we need to set some metric to indicate whether allowed or denied(same as resp_code in http filter) into metadata if mode==shadow.
2.check if effective_policy_id is empty before setting val to metadata ?

Reply:
2. The `effective_policy_id` empty check happens within `setDynamicMetadata` currently.

Reply:
Addressed in 4307545.

Reply:
Lets not do any writes to the dynamic metadata now.. More like, lets do that in separate PR as there is an ongoing discussion on how ot manage the state emitted by different filters. If RBAC starts writing into this state, then things getmore complicated on that front. I would like to sort that discussion with other envoy maintainers.

So, in the interest of time, I suggest scoping this PR purely in terms of bring TCP on parity with HTTP - where we allow custom attributes for both.